### PR TITLE
docs: Use intersphinx to map old versions and cleanup version history

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Use intersphinx to map old versions and cleanup version history
Additional Description:

This PR 

- adds a patched version of intersphinx (~from https://github.com/sphinx-doc/sphinx/pull/8981)
- uses versioned refs for version history
- cleans up `inline literals` in version history

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: docs/root/version_history/v1.10.0.rst [hunk 1]
```diff
@@ -5,97 +5,97 @@ Changes
...
-* config: removed REST_LEGACY as a valid :ref:`ApiType <envoy_api_field_core.ApiConfigSource.api_type>`.
+* config: removed deprecated_v1 sds_config from :ref:`Bootstrap config <v1.10.0:config_overview_v2_bootstrap>`.
+* config: removed the deprecated_v1 config option from :ref:`ring hash <v1.10.0:envoy_api_msg_Cluster.RingHashLbConfig>`.
+* config: removed REST_LEGACY as a valid :ref:`ApiType <v1.10.0:envoy_api_field_core.ApiConfigSource.api_type>`.
 * config: finish cluster warming only when a named response i.e. ClusterLoadAssignment associated to the cluster being warmed comes in the EDS response. This is a behavioural change from the current implementation where warming of cluster completes on missing load assignments also.
 * config: use Envoy cpuset size to set the default number or worker threads if :option:`--cpuset-threads` is enabled.
-* config: added support for :ref:`initial_fetch_timeout <envoy_api_field_core.ConfigSource.initial_fetch_timeout>`. The timeout is disabled by default.
-* cors: added :ref:`filter_enabled & shadow_enabled RuntimeFractionalPercent flags <cors-runtime>` to filter.
-* csrf: added :ref:`CSRF filter <config_http_filters_csrf>`.
```
Comment (phlax) [inline docs/root/version_history/v1.10.0.rst:29]: there is no documented link anchors for csrf in v1.10.0 (and not really any docs - although there is a very vague mention)

File: docs/root/version_history/v1.10.0.rst [hunk 2]
```diff
@@ -5,97 +5,97 @@ Changes
...
-* mysql: added a MySQL proxy filter that is capable of parsing SQL queries over MySQL wire protocol. Refer to :ref:`MySQL proxy<config_network_filters_mysql_proxy>` for more details.
+* outlier_detection: added support for :ref:`outlier detection event protobuf-based logging <v1.10.0:arch_overview_outlier_detection_logging>`.
+* mysql: added a MySQL proxy filter that is capable of parsing SQL queries over MySQL wire protocol. Refer to :ref:`MySQL proxy <v1.10.0:config_network_filters_mysql_proxy>` for more details.
 * performance: new buffer implementation (disabled by default; to test it, add "--use-libevent-buffers 0" to the command-line arguments when starting Envoy).
-* jwt_authn: added :ref:`filter_state_rules <envoy_api_field_config.filter.http.jwt_authn.v2alpha.JwtAuthentication.filter_state_rules>` to allow specifying requirements from filterState by other filters.
+* jwt_authn: added :ref:`filter_state_rules <v1.10.0:envoy_api_field_config.filter.http.jwt_authn.v2alpha.jwtauthentication.rules>` to allow specifying requirements from filterState by other filters.
```
Comment (phlax) [inline docs/root/version_history/v1.10.0.rst:58]: no `filter_state_rules` - but i guess `rules` is the intended link

File: docs/root/version_history/v1.11.0.rst [hunk 3]
```diff
@@ -6,115 +6,115 @@ Changes
...
+* admin: extend :ref:`/runtime_modify endpoint <v1.11.0:operations_admin_interface_runtime_modify>` to support parameters within the request body.
+* admin: the :ref:`/listener endpoint <v1.11.0:operations_admin_interface_listeners>` now returns :ref:`listeners.proto <v1.11.0:envoy_api_msg_admin.v2alpha.Listeners>` which includes listener names and ports.
 * admin: added host priority to :http:get:`/clusters` and :http:get:`/clusters?format=json` endpoint response
-* admin: the :ref:`/clusters endpoint <operations_admin_interface_clusters>` now shows hostname
+* admin: the :ref:`/clusters endpoint <v1.11.0:operations_admin_interface_clusters>` now shows hostname
   for each host, useful for DNS based clusters.
 * api: track and report requests issued since last load report.
 * build: releases are built with Clang and linked with LLD.
-* config: added :ref:stats_server_version_override` <envoy_api_field_config.bootstrap.v2.Bootstrap.stats_server_version_override>` in bootstrap, that can be used to override :ref:`server.version statistic <server_statistics>`.
```
Comment (phlax) [inline docs/root/version_history/v1.11.0.rst:20]: no `stats_server_version_override` - replaced with a link to `stats_config`

Comment (phlax) [reply docs/root/version_history/v1.11.0.rst:20]: also - no `server_statistics` - replaced with a link to just `statistics`

File: docs/root/version_history/v1.13.0.rst [hunk 4]
```diff
@@ -4,82 +4,82 @@
...
+* access log: added FILTER_STATE :ref:`access log formatters <v1.13.0:config_access_log_format>` and gRPC access logger.
+* admin: added the ability to filter :ref:`/config_dump <v1.13.0:operations_admin_interface_config_dump>`.
+* access log: added a :ref:`typed JSON logging mode <v1.13.0:config_access_log_format_dictionaries>` to output access logs in JSON format with non-string values
+* access log: fixed UPSTREAM_LOCAL_ADDRESS :ref:`access log formatters <v1.13.0:config_access_log_format>` to work for http requests
 * access log: added HOSTNAME.
 * api: remove all support for v1
-* api: added ability to specify `mode` for :ref:`Pipe <envoy_api_field_core.Pipe.mode>`.
+* api: added ability to specify `mode` for :ref:`Pipe <v1.13.0:envoy_api_field_core.Pipe.mode>`.
 * api: support for the v3 xDS API added. See :ref:`api_supported_versions`.
-* aws_request_signing: added new alpha :ref:`HTTP AWS request signing filter <config_http_filters_aws_request_signing>`.
```
Comment (phlax) [inline docs/root/version_history/v1.13.0.rst:15]: afaict this was added erroneously to the v1.13 changelog when v1.14 was current - consequently the link doesnt work in v1.13

ive removed the link - the other alternative is to move this to the v1.14 changelog

**Comment by phlax** [comment]
need to investigate an issue with intersphinx...

**Comment by phlax** [comment]
it seems that intersphinx has a very unhelpful "feature" (or > 6-year-old bug depending on pov https://github.com/sphinx-doc/sphinx/issues/2068) by which it kinda guesses which set of docs a link belongs to.

this means that by enabling intersphinx any broken links in current docs will happily pass if it can find the link in *any* of the linked documentation

@nijel from weblate has submitted a patch to fix this (here - https://github.com/sphinx-doc/sphinx/pull/8981) but afaict the sphinx devs either dont understand this bug (and how it makes intersphinx basically unusable) or dont care too much (they are `+-0` to fixing ?) and it wont be fixed until at least sphinx==4.1

for these reasons i have included a patched version of intersphinx with  @nijel 's fix

**Comment by nijel** [comment]
Please comment on the pull request so that Sphinx developers can see there are more people wanting this ;-)

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/16155/docs/index.html

**Comment by htuch** [comment]
@phlax what do you think the likelihood of getting this fixed upstream and landed soon is? I'm reluctant to take on this fork, but I also understand the concern.

**Comment by phlax** [comment]
> ... I also understand the concern.

Yep - i think we definitely want to use this and it definitely has this bug/issue

That means we have to choose the least bad option. The ones i considered

- live with the bug and hope we dont accidentally add/remove link targets
- build the docs twice - first removing the extension and all version info, and then readding it
- create a separate intersphinx package somewhere
- dynamically patch the module during docs build - ie `curl ... > _ext/intersphinx.py ... patch ...`
- copy the module into the repo and just add the fix

of those options the last seems the most straightforward - its ultimately a single module/file that had to be copied in

i can add an Envoy ticket to remind us to remove it and use upstream once the issue is fixed. We may have to update it anyway if we upgrade to sphinx 4

> what do you think the likelihood of getting this fixed upstream and landed soon is?

sphinx 4 is currently in beta - not sure of the timescales to release. The PR for this was bumped 4 -> 4.1 - not sure why - it could just be that it needs rebasing (there were some very minor nits from my reading - which may/not be valid - tbh i dont understand why it wasnt just landed at the time)

**Comment by phlax** [comment]
relatedly, ive begun testing the sphinx4 upgrade 

atm i blocks on sphinx-tabs - there is a PR with compat here https://github.com/executablebooks/sphinx-tabs/pull/110 - so afaict once sphinx4 is out of beta that should unblock

File: docs/_ext/intersphinx_custom.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
+"""
```
Comment (htuch) [inline docs/_ext/intersphinx_custom.py:1]: @phlax can you describe what your changes are at the top of this comment block? And link to an issue filed to track removing?

**Review by htuch — COMMENTED** [review]
LGTM modulo comment.

File: docs/_ext/intersphinx_custom.py (hunk 5 above)
Comment (phlax) [reply docs/_ext/intersphinx_custom.py:1]: yep for sure...

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@phlax needs format fix.

**Comment by phlax** [comment]
seems to have format issues - i will fix tomorrow

/wait

**Review by htuch — APPROVED** [review]

**Pull request merged**